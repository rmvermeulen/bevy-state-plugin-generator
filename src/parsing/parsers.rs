use nom::branch::alt;
use nom::bytes::complete::*;
use nom::character::complete::*;
use nom::combinator::recognize;
use nom::multi::many0;
use nom::sequence::*;
use nom::{IResult, Parser};

use crate::parsing::{Comment, Identifier, ParseNode, Token};

pub fn parse_comment(input: &str) -> IResult<&str, ParseNode<'_>> {
    comment(input).map_result(ParseNode::Comment)
}

pub fn comment(input: &str) -> IResult<&str, Comment<'_>> {
    use nom::combinator::eof;

    delimited(skip_to(tag("//")), not_line_ending, alt((eof, line_ending)))
        .parse(input)
        .map_result(|c| c.trim().into())
}

pub fn identifier(input: &str) -> IResult<&str, Identifier<'_>> {
    recognize(pair(
        take_till1(|c: char| !c.is_uppercase()),
        take_while(|c: char| c.is_alphanumeric() || c == '_'),
    ))
    .parse(input)
    .map_result(Identifier::from)
}

fn skip_to<P, T, O, E>(parser: P) -> impl Parser<T, Output = O, Error = E>
where
    T: nom::Input,
    P: Parser<T, Output = O, Error = E>,
    <T as nom::Input>::Item: nom::AsChar,
    E: nom::error::ParseError<T>,
{
    preceded(multispace0, parser)
}

pub fn separator(input: &str) -> IResult<&str, Token> {
    skip_to(tag(","))
        .parse(input)
        .map_result(|_| Token::Separator)
}

pub fn open_enum(input: &str) -> IResult<&str, Token> {
    skip_to(tag("{"))
        .parse(input)
        .map_result(|_| Token::OpenEnum)
}

pub fn close_enum(input: &str) -> IResult<&str, Token> {
    skip_to(tag("}"))
        .parse(input)
        .map_result(|_| Token::CloseEnum)
}

pub fn open_list(input: &str) -> IResult<&str, Token> {
    skip_to(tag("["))
        .parse(input)
        .map_result(|_| Token::OpenList)
}

pub fn close_list(input: &str) -> IResult<&str, Token> {
    skip_to(tag("]"))
        .parse(input)
        .map_result(|_| Token::CloseList)
}

pub trait MapResult<'a, I, O1, O2> {
    fn map_result(self, callback: impl Fn(O1) -> O2) -> IResult<I, O2>;
}

impl<I, O1, O2> MapResult<'_, I, O1, O2> for IResult<I, O1> {
    fn map_result(self, callback: impl Fn(O1) -> O2) -> IResult<I, O2> {
        self.map(|(rest, result)| (rest, callback(result)))
    }
}

pub fn parse_config(input: &str) -> IResult<&str, Vec<ParseNode<'_>>> {
    many0(delimited(many0(separator), parse_node, many0(separator))).parse(input)
}

pub fn parse_node(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    alt((parse_enum, parse_list, parse_comment, parse_singleton)).parse(input)
}

pub fn parse_singleton(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    skip_to(identifier)
        .parse(input)
        .map_result(ParseNode::singleton)
}

pub fn parse_enum(input: &str) -> IResult<&str, ParseNode<'_>> {
    let (input, name) = terminated(skip_to(identifier), skip_to(open_enum)).parse(input)?;
    let (input, children) = skip_to(parse_elements_until(close_enum)).parse(input)?;
    Ok((input, ParseNode::Enum(name, children)))
}

pub fn parse_list(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    let (input, name) = skip_to(identifier).parse(input)?;
    let (input, children) =
        skip_to(preceded(open_list, parse_elements_until(close_list))).parse(input)?;
    Ok((input, ParseNode::List(name, children)))
}

pub fn parse_elements_until<'a>(
    until: impl Fn(&'a str) -> IResult<&'a str, Token> + Copy,
) -> impl Fn(&'a str) -> IResult<&'a str, Vec<ParseNode<'a>>> {
    move |input: &'a str| {
        delimited(
            // ignore any leading whitespace and commas
            many0(separator),
            // 0 or more elements, ignoring whitespace and commas
            many0(alt((terminated(parse_node, many0(separator)),))),
            // then expect the closing token
            until,
        )
        .parse(input)
    }
}
