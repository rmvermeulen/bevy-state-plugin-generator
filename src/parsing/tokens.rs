use derive_more::{Deref, From};

#[derive(Debug, PartialEq)]
pub enum Token {
    Separator,
    OpenEnum,
    CloseEnum,
    OpenList,
    CloseList,
}

#[derive(Clone, Copy, Debug, Deref, From, PartialEq)]
pub struct Identifier<'a>(&'a str);

impl std::fmt::Display for Identifier<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Clone, Copy, Debug, Deref, From, PartialEq)]
pub struct Comment<'a>(&'a str);

#[derive(Clone, Debug, PartialEq)]
pub enum ParseNode<'a> {
    Singleton(Identifier<'a>),
    Enum(Identifier<'a>, Vec<ParseNode<'a>>),
    List(Identifier<'a>, Vec<ParseNode<'a>>),
    Comment(Comment<'a>),
}

impl<'a> ParseNode<'a> {
    pub fn singleton<I: Into<Identifier<'a>>>(name: I) -> Self {
        Self::Singleton(name.into())
    }
    fn identifier(&self) -> Option<&Identifier<'a>> {
        match self {
            Self::Singleton(id) | Self::Enum(id, _) | Self::List(id, _) => Some(id),
            _ => None,
        }
    }
    pub fn name(&self) -> Option<&str> {
        self.identifier().map(|id| id.0)
    }
    pub fn children(&self) -> Vec<ParseNode<'a>> {
        match self {
            Self::Singleton(_) | Self::Comment(_) => Vec::new(),
            Self::Enum(_, children) => children.clone(),
            Self::List(_, children) => children.clone(),
        }
    }
    pub fn get_tree_size(&self) -> usize {
        match self {
            Self::Comment(_) => 1,
            Self::Singleton(_) => 1,
            Self::Enum(_, children) | Self::List(_, children) => {
                children
                    .iter()
                    .map(|child| child.get_tree_size())
                    .sum::<usize>()
                    + 1
            }
        }
    }
}

impl<'a> TryFrom<&'a str> for ParseNode<'a> {
    type Error = nom::Err<nom::error::Error<&'a str>>;

    fn try_from(s: &'a str) -> Result<Self, Self::Error> {
        crate::parsing::parse_node(s).map(|(_, node)| node)
    }
}

/// constructors used for testing
#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
impl<'a> ParseNode<'a> {
    pub fn enumeration<I: Into<Identifier<'a>>, V: IntoIterator<Item = ParseNode<'a>>>(
        name: I,
        variants: V,
    ) -> Self {
        Self::Enum(name.into(), variants.into_iter().collect())
    }
    pub fn list_empty<I: Into<Identifier<'a>>>(name: I) -> Self {
        Self::List(name.into(), vec![])
    }
    pub fn list<I: Into<Identifier<'a>>, V: IntoIterator<Item = ParseNode<'a>>>(
        name: I,
        variants: V,
    ) -> Self {
        Self::List(name.into(), variants.into_iter().collect())
    }
    pub fn comment<C: Into<Comment<'a>>>(name: C) -> Self {
        Self::Comment(name.into())
    }
}
