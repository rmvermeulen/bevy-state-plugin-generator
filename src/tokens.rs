use derive_more::{Deref, From};

#[derive(Debug, PartialEq)]
pub enum Token {
    Separator,
    OpenEnum,
    CloseEnum,
    OpenList,
    CloseList,
}

#[derive(Debug, Deref, From, PartialEq)]
pub struct Identifier<'a>(&'a str);

impl std::fmt::Display for Identifier<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug, Deref, From, PartialEq)]
pub struct Comment<'a>(&'a str);

#[derive(Debug, PartialEq)]
pub enum ParseNode<'a> {
    Singleton(Identifier<'a>),
    Enum(Identifier<'a>, Vec<ParseNode<'a>>),
    List(Identifier<'a>, Vec<ParseNode<'a>>),
    Comment(Comment<'a>),
}

impl<'a> ParseNode<'a> {
    pub fn singleton<I: Into<Identifier<'a>>>(name: I) -> Self {
        Self::Singleton(name.into())
    }
    pub fn enumeration<I: Into<Identifier<'a>>, V: IntoIterator<Item = ParseNode<'a>>>(
        name: I,
        variants: V,
    ) -> Self {
        Self::Enum(name.into(), variants.into_iter().collect())
    }
    #[cfg(test)]
    pub fn list_empty<I: Into<Identifier<'a>>>(name: I) -> Self {
        Self::List(name.into(), vec![])
    }
    #[cfg(test)]
    pub fn list<I: Into<Identifier<'a>>, V: IntoIterator<Item = ParseNode<'a>>>(
        name: I,
        variants: V,
    ) -> Self {
        Self::List(name.into(), variants.into_iter().collect())
    }
    #[cfg(test)]
    pub fn comment<C: Into<Comment<'a>>>(name: C) -> Self {
        Self::Comment(name.into())
    }
}

impl<'a> TryFrom<&'a str> for ParseNode<'a> {
    type Error = nom::Err<nom::error::Error<&'a str>>;

    fn try_from(s: &'a str) -> Result<Self, Self::Error> {
        crate::parsing::parse_node(s).map(|(_, node)| node)
    }
}
