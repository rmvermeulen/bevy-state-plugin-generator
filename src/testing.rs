pub use insta::*;
pub use rstest::*;
pub use speculoos::prelude::*;

pub use crate::set_snapshot_suffix;

#[macro_export]
#[allow(missing_docs)]
macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

pub mod node_data {
    use bevy_utils::default;

    use super::*;
    use crate::processing::NodeData;

    fn node<S: ToString>(name: S, index: usize, depth: usize, parent: Option<usize>) -> NodeData {
        let name = name.to_string();
        NodeData {
            index,
            parent,
            resolved_name: Some(name.clone()),
            name,
            depth,
            ..default()
        }
    }

    #[fixture]
    pub fn single_node() -> Vec<NodeData> {
        vec![node("RootState", 0, 0, None)]
    }

    #[fixture]
    pub fn nested_example() -> Vec<NodeData> {
        vec![
            node("RootState", 0, 0, None),
            node("Menu", 1, 1, Some(0)),
            node("Options", 2, 2, Some(1)),
            node("Audio", 3, 3, Some(2)),
            node("Video", 4, 3, Some(2)),
        ]
    }
}

pub mod parse_node {
    use super::*;
    use crate::parsing::Node;

    #[fixture]
    pub fn duplicate_names() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("A")])
    }

    #[fixture]
    pub fn overlapping_names() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::enumeration("Root", [Node::singleton("A")]),
                Node::singleton("A"),
            ],
        )
    }

    #[fixture]
    pub fn enum_root_a() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn enum_root_ab() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn enum_root_a_b() -> Node<'static> {
        Node::enumeration("Root", [Node::enumeration("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn enum_root_a_b_up_c() -> Node<'static> {
        Node::enumeration(
            "Root",
            [
                Node::enumeration("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn list_root_a() -> Node<'static> {
        Node::list("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn list_root_ab() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn list_root_a_b() -> Node<'static> {
        Node::list("Root", [Node::list("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn list_root_a_b_up_c() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::list("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn nested_example() -> Node<'static> {
        Node::enumeration(
            "Menu",
            [
                Node::singleton("Main"),
                Node::comment("these are the options"),
                Node::enumeration(
                    "Options",
                    [
                        Node::singleton("Graphics"),
                        Node::singleton("Audio"),
                        Node::singleton("Gameplay"),
                    ],
                ),
                Node::enumeration("Game", [Node::singleton("Save"), Node::singleton("Load")]),
            ],
        )
    }
}
