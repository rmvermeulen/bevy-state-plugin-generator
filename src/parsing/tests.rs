use nom::Parser;
use nom::branch::alt;

use super::parsers::*;
use crate::testing::*;
use crate::tokens::*;

#[rstest]
#[case("Name", "Name")]
#[case("Name ", "Name")]
#[case("Name123", "Name123")]
#[case("Name_123", "Name_123")]
fn test_identifier(#[case] input: &str, #[case] token: &str) {
    assert_that!(identifier(input))
        .is_ok()
        .map(|(_, token)| token)
        .is_equal_to(Identifier::from(token));
}

#[rstest]
#[case("//\nHello\n", "")]
#[case("//Hello\n", "Hello")]
#[case("// Hello\n", "Hello")]
#[case("// Hello// \n", "Hello//")]
#[case("// Hello, how are you? \n", "Hello, how are you?")]
fn test_comment(#[case] input: &str, #[case] expected: &str) {
    assert_that!(comment(input))
        .is_ok()
        .map(|(_, token)| token)
        .is_equal_to(Comment::from(expected));
}

#[rstest]
#[case("//Hello\n", "Hello")]
#[case("// Hello\n", "Hello")]
#[case("// Hello// \n", "Hello//")]
#[case("// Hello, how are you? \n", "Hello, how are you?")]
fn test_parse_comment(#[case] input: &str, #[case] expected: &str) {
    assert_that!(parse_comment(input))
        .is_ok()
        .map(|(_, token)| token)
        .is_equal_to(ParseNode::comment(expected));
}

#[rstest]
#[case("{", Token::OpenEnum)]
#[case("}", Token::CloseEnum)]
#[case("[", Token::OpenList)]
#[case("]", Token::CloseList)]
fn test_single_char_tokens(#[case] input: &str, #[case] expected: Token) {
    let mut parser = alt((open_enum, close_enum, open_list, close_list));

    assert_that!(parser.parse(input))
        .is_ok()
        .is_equal_to(("", expected));
}

#[rstest]
#[case("Name,", ",", ParseNode::singleton("Name"))]
#[case("  Name,", ",", ParseNode::singleton("Name"))]
#[case("First, Second", ", Second", ParseNode::singleton("First"))]
fn test_parse_singleton(#[case] input: &str, #[case] rest: &str, #[case] node: ParseNode) {
    assert_that!(parse_singleton(input).unwrap()).is_equal_to((rest, node));
}

#[rstest]
fn test_parse_enum_empty() {
    assert_compact_debug_snapshot!(parse_enum("Name{}").unwrap(), @r#"("", Enum(Identifier("Name"), []))"#);
    assert_compact_debug_snapshot!(parse_enum("Name {}").unwrap(), @r#"("", Enum(Identifier("Name"), []))"#);
}

#[rstest]
#[case("Name {A}",  ParseNode::enumeration("Name", [ParseNode::singleton("A")]))]
#[case("Name { A}",  ParseNode::enumeration("Name", [ParseNode::singleton("A")]))]
#[case("Name {A }",  ParseNode::enumeration("Name", [ParseNode::singleton("A")]))]
#[case("Name { A }",  ParseNode::enumeration("Name", [ParseNode::singleton("A")]))]
fn test_parse_enum_single(#[case] input: &str, #[case] node: ParseNode) {
    assert_that!(parse_enum(input).unwrap()).is_equal_to(("", node));
}

#[rstest]
#[case("Name {A,B}", ParseNode::enumeration("Name", [
    ParseNode::singleton("A"),
    ParseNode::singleton("B")
]))]
#[case("Name { A B }",  ParseNode::enumeration("Name", [
    ParseNode::singleton("A"),
    ParseNode::singleton("B")
]))]
#[case("Name { A { B } }",  ParseNode::enumeration("Name", [
    ParseNode::enumeration("A", [ParseNode::singleton("B")])
]))]
#[case("Name { A { B }, C }",  ParseNode::enumeration("Name", [
    ParseNode::enumeration("A", [ParseNode::singleton("B")]),
    ParseNode::singleton("C")
]))]
fn test_parse_enum_variants(#[case] input: &str, #[case] node: ParseNode) {
    assert_that!(parse_enum(input).unwrap()).is_equal_to(("", node));
}

#[rstest]
#[case::just_a_comma("Name {,}", ParseNode::enumeration("Name", [ ]))]
#[case::mora_commas("Name {,,,,}", ParseNode::enumeration("Name", [ ]))]
#[case::comma_after_variant("Name {A,}",
    ParseNode::enumeration("Name", [ ParseNode::singleton("A") ]))]
#[case::comma_before_variant("Name {,A}",
    ParseNode::enumeration("Name", [ ParseNode::singleton("A") ]))]
#[case::comma_between_variants("Name {A,B}",
    ParseNode::enumeration("Name", [
        ParseNode::singleton("A"),
        ParseNode::singleton("B")
    ])
)]
fn test_parse_enum_optional_commas(#[case] input: &str, #[case] node: ParseNode) {
    assert_that!(parse_enum(input).unwrap()).is_equal_to(("", node));
}

#[rstest]
#[case("Name []", ParseNode::list_empty("Name"))]
#[case("Name[]", ParseNode::list_empty("Name"))]
#[case("Name[A]",  ParseNode::list("Name", [ParseNode::singleton("A")]))]
#[case("Name[A,B]",  ParseNode::list("Name", [
    ParseNode::singleton("A"),
    ParseNode::singleton("B"),
]))]
fn test_parse_list(#[case] input: &str, #[case] node: ParseNode) {
    assert_that!(parse_list(input).unwrap()).is_equal_to(("", node));
}

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

#[rstest]
#[case("Name")]
#[case("Name {A}")]
fn test_parse_node(#[case] input: &str) {
    set_snapshot_suffix!("{}", input.replace(" ", "_"));
    assert_compact_debug_snapshot!(parse_node(input));
}

#[rstest]
fn test_parse_node_nested_enums() {
    let input = "Name { A { B, C {D E {F G}} H } I J }";
    assert_debug_snapshot!(parse_node(input));
}

#[rstest]
#[case("//Comment", ParseNode::comment("Comment"))]
#[case("// Comment", ParseNode::comment("Comment"))]
fn test_parse_node_with_comments(#[case] input: &str, #[case] comment: ParseNode) {
    assert_that!(parse_node(input)).is_ok_containing(("", comment));
}

#[rstest]
fn test_parse_node_messy_example() {
    let input = "Name [ A { B, C [D E {F G}] H } I J ]";
    assert_debug_snapshot!(parse_node(input));
}

#[rstest]
fn test_parse_list_incomplete() {
    assert_compact_debug_snapshot!(parse_list("Name [ A"), @r#"Err(Error(Error { input: "", code: Tag })) "#);
}

#[rstest]
fn test_parse_enum_incomplete() {
    assert_compact_debug_snapshot!(parse_enum("Name { A"), @r#"Err(Error(Error { input: "", code: Tag }))"#);
}

#[rstest]
#[case("A", vec![ParseNode::singleton("A")])]
#[case("A B", vec![ParseNode::singleton("A"), ParseNode::singleton("B")] )]
#[case("A,B", vec![ParseNode::singleton("A"), ParseNode::singleton("B")] )]
#[case("A, B", vec![ParseNode::singleton("A"), ParseNode::singleton("B")] )]
#[case("A{B}", vec![ParseNode::enumeration("A", [ ParseNode::singleton("B") ])] )]
#[case("A{B,C D,}", vec![ParseNode::enumeration("A", [
    ParseNode::singleton("B"),
    ParseNode::singleton("C"),
    ParseNode::singleton("D")
])])]
#[case("A{B{C{D{E{F}}}}}", vec![ParseNode::enumeration("A", [
    ParseNode::enumeration("B", [
        ParseNode::enumeration("C", [
            ParseNode::enumeration("D", [
                ParseNode::enumeration("E", [
                    ParseNode::singleton("F")
                ])
            ])
        ])
    ])
])])]
#[ case("//A//{ B C }", vec![ ParseNode::comment("A//{ B C }") ])]
#[ case("A//{ B C }", vec![
    ParseNode::singleton("A"),
    ParseNode::comment("{ B C }")
])]
fn test_parse_config(#[case] input: &str, #[case] expected: Vec<ParseNode>) {
    assert_that!(parse_config(input))
        .named(&format!("\"{input}\""))
        .is_ok()
        .is_equal_to(("", expected));
}

#[rstest]
#[case("A//\n{ B C }", vec![
    ParseNode::singleton("A"),
    ParseNode::comment("")
], "{ B C }")]
#[case("A {\n // B\n C\n}...", vec![ ParseNode::enumeration("A", [
    ParseNode::comment("B"),
    ParseNode::singleton("C")
]) ], "...")]
fn test_parse_config_incomplete(
    #[case] input: &str,
    #[case] expected: Vec<ParseNode>,
    #[case] rest: &str,
) {
    assert_that!(parse_config(input))
        .named(&format!("\"{input}\""))
        .is_ok()
        .is_equal_to((rest, expected));
}

#[rstest]
#[case("Name, Name2", vec![
    ParseNode::singleton("Name"),
    ParseNode::singleton("Name2"),
])]
fn test_parse_states_file(#[case] input: &str, #[case] expected: Vec<ParseNode>) {
    assert_that!(parse_states_text(input))
        .is_ok()
        .is_equal_to(expected);
}
