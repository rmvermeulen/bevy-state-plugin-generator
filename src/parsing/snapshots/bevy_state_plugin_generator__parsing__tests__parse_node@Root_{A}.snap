---
source: src/parsing/tests.rs
expression: parse_node(input)
---
Ok(("", Enum(Identifier("Root"), [Singleton(Identifier("A"))])))
